File: Classes/SwitchTableCell.m
// SwitchTableCell.m
//
// Kraftstoff


#import "SwitchTableCell.h"
#import "AppDelegate.h"

static CGFloat const margin = 8.0;


@implementation SwitchTableCell

- (void)finishConstruction
{
	[super finishConstruction];

    // No highlight on touch
    self.selectionStyle = UITableViewCellSelectionStyleNone;

    // Create switch
    self.valueSwitch = [[UISwitch alloc] initWithFrame:CGRectZero];
    [self.valueSwitch addTarget:self action:@selector(switchToggledAction:) forControlEvents:UIControlEventValueChanged];

	[self.contentView addSubview:self.valueSwitch];

    // Configure the alternate textlabel
    self.valueLabel = [[UILabel alloc] initWithFrame:CGRectZero];

    self.valueLabel.font             = [UIFont fontWithName:@"HelveticaNeue-Light" size:17.0];
	self.valueLabel.textAlignment    = NSTextAlignmentRight;
	self.valueLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth;
	self.valueLabel.backgroundColor  = [UIColor clearColor];
	self.valueLabel.textColor        = [UIColor blackColor];

    self.valueLabel.hidden                 = YES;
	self.valueLabel.userInteractionEnabled = NO;

    [self.contentView addSubview:self.valueLabel];

    // Configure the default textlabel
    UILabel *label = self.textLabel;

	label.textAlignment        = NSTextAlignmentLeft;
	label.font                 = [UIFont fontWithName:@"HelveticaNeue" size:17.0];
	label.highlightedTextColor = [UIColor blackColor];
	label.textColor            = [UIColor blackColor];
}


- (void)configureForData:(id)dataObject
          viewController:(id)viewController
               tableView:(UITableView *)tableView
               indexPath:(NSIndexPath *)indexPath
{
	[super configureForData:dataObject viewController:viewController tableView:tableView indexPath:indexPath];

	self.textLabel.text   = ((NSDictionary *)dataObject)[@"label"];
    self.delegate         = viewController;
    self.valueIdentifier  = ((NSDictionary *)dataObject)[@"valueIdentifier"];

    BOOL isON = [[self.delegate valueForIdentifier:self.valueIdentifier] boolValue];

    [self.valueSwitch setOn:isON];
    [self.valueLabel setText:NSLocalizedString(isON ? @"Yes" : @"No", @"")];

    BOOL showAlternate = [[self.delegate valueForIdentifier:@"showValueLabel"] boolValue];

    self.valueSwitch.hidden =  showAlternate;
    self.valueLabel.hidden  = !showAlternate;
}


- (void)layoutSubviews
{
    [super layoutSubviews];


    CGFloat leftOffset = 6.0;

    // Text label on the left
    CGFloat labelWidth = [self.textLabel.text sizeWithAttributes:@{NSFontAttributeName:self.textLabel.font}].width;
    CGFloat height     = self.contentView.bounds.size.height;
	CGFloat width      = self.contentView.bounds.size.width;

    self.textLabel.frame = CGRectMake (margin + leftOffset, 0.0, labelWidth, height - 1);

    // UISwitch
    CGRect valueFrame = self.valueSwitch.frame;
    self.valueSwitch.frame = CGRectMake (width - margin - valueFrame.size.width,
                                         floor ((height - valueFrame.size.height)/2),
                                         valueFrame.size.width,
                                         valueFrame.size.height);

    // Alternate for UISwitch
    CGFloat alternateHeight = [self.valueLabel.text sizeWithAttributes:@{NSFontAttributeName:self.valueLabel.font}].height;
    self.valueLabel.frame = CGRectMake (width - margin - 100.0, floor ((height - alternateHeight)/2), 100.0, alternateHeight);
}


- (void)switchToggledAction:(UISwitch*)sender
{
    BOOL isON = [sender isOn];

    [self.delegate valueChanged:@(isON) identifier:self.valueIdentifier];
    [self.valueLabel setText:NSLocalizedString(isON ? @"Yes" : @"No", @"")];
}

@end

File: Classes/SwitchTableCell.swift
//
//  SwitchTableCell.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 03.05.15.
//
//

import UIKit

final class SwitchTableCell: PageCell {
  private let margin: CGFloat = 8.0

  var keyLabel: UILabel
  var valueSwitch: UISwitch
  var valueLabel: UILabel
  var valueIdentifier: String!

  weak var delegate: EditablePageCellDelegate?

  required init() {
    keyLabel = UILabel(frame: .zero)
    valueSwitch = UISwitch(frame: .zero)
    valueLabel = UILabel(frame: .zero)

    super.init()

    // No highlight on touch
    selectionStyle = .none

    // Create switch
    valueSwitch.addTarget(self, action: #selector(SwitchTableCell.switchToggledAction(_:)), for: UIControl.Event.valueChanged)
    valueSwitch.translatesAutoresizingMaskIntoConstraints = false

    contentView.addSubview(valueSwitch)

    // Configure the alternate textlabel

    valueLabel.textAlignment = .right
    valueLabel.backgroundColor = .clear
    valueLabel.textColor = .label
    valueLabel.isHidden = true
    valueLabel.isUserInteractionEnabled = false
    valueLabel.translatesAutoresizingMaskIntoConstraints = false
    valueLabel.adjustsFontForContentSizeCategory = true
    valueLabel.font = UIFont.preferredFont(forTextStyle: UIFont.TextStyle.body)

    contentView.addSubview(valueLabel)

    // Configure the default textlabel
    keyLabel.textAlignment = .left
    keyLabel.highlightedTextColor = .label
    keyLabel.textColor = .label
    keyLabel.translatesAutoresizingMaskIntoConstraints = false
    keyLabel.adjustsFontForContentSizeCategory = true
    keyLabel.font = UIFont.preferredFont(forTextStyle: UIFont.TextStyle.body)

    contentView.addSubview(keyLabel)

    let constraints = Array([
      [
        NSLayoutConstraint(item: valueSwitch, attribute: .centerY, relatedBy: .equal, toItem: contentView, attribute: .centerY, multiplier: 1.0, constant: 0.0),
        NSLayoutConstraint(item: valueLabel, attribute: .centerY, relatedBy: .equal, toItem: contentView, attribute: .centerY, multiplier: 1.0, constant: 0.0),
      ],
      NSLayoutConstraint.constraints(withVisualFormat: "|-[keyLabel]-[valueSwitch]-|", options: [], metrics: nil, views: ["keyLabel": keyLabel, "valueSwitch": valueSwitch]),
      NSLayoutConstraint.constraints(withVisualFormat: "|-[keyLabel]-[valueLabel]-|", options: [], metrics: nil, views: ["keyLabel": keyLabel, "valueLabel": valueLabel]),
      NSLayoutConstraint.constraints(withVisualFormat: "V:|-[keyLabel]-|", options: [], metrics: nil, views: ["keyLabel": keyLabel]),
      NSLayoutConstraint.constraints(withVisualFormat: "V:|-(>=4)-[valueSwitch]-(>=4)-|", options: [], metrics: nil, views: ["valueSwitch": valueSwitch]),
    ].joined())
    NSLayoutConstraint.activate(constraints)
  }

  @available(*, unavailable)
  required init(coder _: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func configureForData(_ dictionary: [String: Any], viewController: UIViewController, tableView: UITableView, indexPath: IndexPath) {
    super.configureForData(dictionary, viewController: viewController, tableView: tableView, indexPath: indexPath)

    keyLabel.text = dictionary["label"] as? String
    delegate = viewController as? EditablePageCellDelegate
    valueIdentifier = dictionary["valueIdentifier"] as? String

    let isOn = delegate?.valueForIdentifier(valueIdentifier) as? Bool ?? false

    valueSwitch.isOn = isOn
    valueLabel.text = NSLocalizedString(isOn ? "Yes" : "No", comment: "")

    let showAlternate = delegate?.valueForIdentifier("showValueLabel") as? Bool ?? false

    valueSwitch.isHidden = showAlternate
    valueLabel.isHidden = !showAlternate
  }

  @objc func switchToggledAction(_ sender: UISwitch) {
    let isOn = sender.isOn

    delegate?.valueChanged(isOn, identifier: valueIdentifier)
    valueLabel.text = NSLocalizedString(isOn ? "Yes" : "No", comment: "")
  }

  override func reset() {
    super.reset()

    keyLabel.text = ""
    valueLabel.text = ""
  }
}
